File: Laboratori/Ghigo/Lab0/Consumatore.java
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;

// Consumatore e' un filtro
public class Consumatore {
	public static void main(String[] args) {
		FileReader fileReader = null;
        String regex = args[1];
		char ch;
		int x;
		
		if (args.length != 2){
			System.out.println("Utilizzo: consumatore <inputFilename> <caratteri da rimuovere>");
			System.exit(0);
		}
	  
		try {
			fileReader = new FileReader(args[0]);
		} catch(FileNotFoundException e){
			System.out.println("File non trovato");
			System.exit(1);
		}
		try {

			while ((x = fileReader.read()) >= 0) { 
				ch = (char) x;
                    if(regex.indexOf(ch) == -1){ // se il xitcarattere non è presente nella stringa regex
                        System.out.print(ch);
                    }
			}
			fileReader.close();
		} catch(IOException ex){
			System.out.println("Errore di input");
			System.exit(2);
		}
}}

// done

File: Laboratori/Ghigo/Lab0/consumatore.c
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#define MAX_STRING_LENGTH 256

// consumatore.c e' un filtro
int main(int argc, char *argv[]) {

    char *file_in, *regex_char, read_char, buf[MAX_STRING_LENGTH];
    int   nread, fd;

    // controllo numero argomenti
    if (argc != 3) {
        perror(" numero di argomenti sbagliato, usage: ./consumatore <file di input> <caratteri da rimuovere>");
        exit(1);
    }

    file_in = argv[1]; // nome del file
    regex_char = argv[2]; // caratteri da rimuovere

    fd = open(file_in, O_RDONLY); // apertura file in lettura
    if (fd < 0) {
        perror("P0: Impossibile aprire il file.");
        exit(2);
    }

    // lettura del file
    while (nread = read(fd, &read_char, sizeof(char))) /* Fino ad EOF*/ {
        if (nread > 0){ // se ho letto qualcosa
            //se il carattere letto è nel set di caratteri da rimuovere, non lo stampo
            if (strchr(regex_char, read_char) == NULL){
            putchar(read_char);
            }
        }
        else {
            printf("(PID %d) impossibile leggere dal file %s", getpid(), file_in);
            perror("Errore in lettura");
            close(fd);
            exit(3);
        }
    }
    close(fd);
}

// done
